storage/spanstore/downsampling_writer.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanstore
+
+import (
+	"hash"
+	"hash/fnv"
+	"math"
+	"sync"
+
+	"github.com/uber/jaeger-lib/metrics"
+
+	"github.com/jaegertracing/jaeger/model"
+)
+
+const (
+	traceIDByteSize = 16
+)
+
+// hasher includes data we want to put in sync.Pool
+type hasher struct {
+	fnv       hash.Hash64
+	byteArray []byte
+}
+
+// downsamplingWriterMetrics keeping track of total number of dropped spans and accepted spans
+type downsamplingWriterMetrics struct {
+	SpansDropped  metrics.Counter `metric:"spans_dropped"`
+	SpansAccepted metrics.Counter `metric:"spans_accepted"`
+}
+
+// DownSamplingWriter is a span Writer that drops spans with a predefined downSamplingRatio.
+type DownSamplingWriter struct {
+	spanWriter   Writer
+	threshold    uint64
+	lengthOfSalt int
+	pool         *sync.Pool
+	metrics      *downsamplingWriterMetrics
```

Comment:
don't need pointer

Reply:
Just curious, how do we decide if metrics or more general struct should have a pointer or not? I saw a lot of examples in internal version with pointer.

Reply:
@yurishkuro comment on this?

Reply:
There's no hard rule, but from efficiency pov, when you don't need a pointer, don't create one.